### ISSUE TITLE
`items_after_statements` cannot be allowed if the item is defined in a macro

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#10540]
### Summary

I have an item defined in a macro. The fact it uses an item is an implementation detail, so I want to ignore this lint. Adding `#[allow(clippy::items_after_statements)]` has no effect.

### Reproducer

I tried this code:

```rust
#![warn(clippy::pedantic)]

macro_rules! static_assert_size {
    ($ty:ty, $size:expr) => {
        #[allow(clippy::items_after_statements)]
        const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
    };
}

pub fn foo() {
    let _ = 1;
    static_assert_size!(u32, 4);
}
```

I expected to see this happen: Clippy emits no warning, since I have an allow above the statement.

Instead, this happened:
```
warning: adding items after statements is confusing, since items exist from the start of the scope
  --> src/lib.rs:6:9
   |
6  |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
12 |     static_assert_size!(u32, 4);
   |     --------------------------- in this macro invocation
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#items_after_statements
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
1  | #![warn(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::items_after_statements)]` implied by `#[warn(clippy::pedantic)]`
   = note: this warning originates in the macro `static_assert_size` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Version

```text
1.70.0-nightly-2023-03-23
```

### Additional Labels

@rustbot label T-macros

**Comment by jyn514** [comment]
Note that putting the allow above the macro invocation doesn't work either:
```
warning: unused attribute `allow`
   --> src/cloud/mod.rs:102:5
    |
102 |     #[allow(clippy::items_after_statements)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: the built-in attribute `allow` will be ignored, since it's applied to the macro invocation `static_assert_size`
   --> src/cloud/mod.rs:103:5
    |
103 |     static_assert_size!(Flow, 456);
    |     ^^^^^^^^^^^^^^^^^^
```
